nntrainer/tensor/manager.cpp

```diff
@@ -245,40 +250,87 @@ void Manager::initialize() {
   bool is_act_layer = layer_type == ActivationLayer::type;
+  bool is_flat_layer = layer_type == FlattenLayer::type;
 
   size_t inout_derivative_size = 0;
 
   std::vector<std::shared_ptr<Var_Grad>> in_out;
-  in_out.reserve(input_dim.size());
+  in_out.reserve(inout_dim.size());
 
-  for (auto const &dim : input_dim) {
-    in_out.emplace_back(
-      std::make_shared<Var_Grad>(dim, true, base_name + std::to_string(cnt++)));
+  for (auto const &dim : inout_dim) {
+    in_out.emplace_back(std::make_shared<Var_Grad>(
+      dim, true, layer_name + std::to_string(cnt++)));
     if (is_act_layer)
       inout_derivative_size += dim.getDataLen();
   }
 
   in_outs.push_back(in_out);
   is_act_type.push_back(is_act_layer);
+  is_flat_type.push_back(is_flat_layer);
 
   max_derivative_size = std::max(max_derivative_size, inout_derivative_size);
   return in_outs.back();
 }
 
+std::vector<std::shared_ptr<Var_Grad>> &
+Manager::trackLayerOutputs(const std::string &layer_type,
+                           const std::string &layer_name,
+                           const std::vector<TensorDim> &output_dim,
+                           const std::vector<TensorDim> &input_dim) {
+  if (enable_inference_inout_memory_opt && input_dim.empty()) {
+    throw std::invalid_argument(
+      "Input dimensions are required with inference memory opt.");
+  } else {
+    size_t shared_inout = 0;
+
+    for (auto const &dim : output_dim)
+      shared_inout += dim.getDataLen();
+
+    for (auto const &dim : input_dim)
+      shared_inout += dim.getDataLen();
+
+    max_shared_inout = std::max(max_shared_inout, shared_inout);
+  }
+
+  return trackLayerInOuts(layer_type, layer_name + ":InOut", output_dim);
+}
+
+std::vector<std::shared_ptr<Var_Grad>> &
+Manager::trackLayerInputs(const std::string &layer_type,
+                          const std::string &layer_name,
+                          const std::vector<TensorDim> &input_dim,
+                          const std::vector<TensorDim> &output_dim) {
+  if (enable_inference_inout_memory_opt && output_dim.empty()) {
+    throw std::invalid_argument(
+      "Output dimensions are required with inference memory opt.");
+  } else {
+    size_t shared_inout = 0;
+
+    for (auto const &dim : output_dim)
+      shared_inout += dim.getDataLen();
+
+    for (auto const &dim : input_dim)
+      shared_inout += dim.getDataLen();
+
+    max_shared_inout = std::max(max_shared_inout, shared_inout);
+  }
+
+  return trackLayerInOuts(layer_type, layer_name + ":InOut", input_dim);
+}
+
 void Manager::untrackLayerInOuts(const std::string layer_name) {
   auto var_name = layer_name + ":InOut" + std::to_string(0);
 
   for (unsigned int cnt = 0; cnt < in_outs.size(); cnt++) {
     if (!in_outs[cnt].empty() && in_outs[cnt][0]->getName() == var_name) {
       in_outs.erase(in_outs.begin() + cnt);
       is_act_type.erase(is_act_type.begin() + cnt);
-      break;
+      is_flat_type.erase(is_flat_type.begin() + cnt);
```

Comment:
Is removing break intended here?

Reply:
Yes, as both input and output currently use the same name when added tracked input/output.
So, it is possible to have two memory allocations with same name, and hence break was removed.

Updated names to be different and added break.